Add range proving

Vimalakirti opened the pull request:
Addressed the issue in https://github.com/uiuc-kang-lab/zk-torch/issues/58

Vimalakirti committed:
add range proving

Vimalakirti commented:
Let me briefly explain the proving idea here. 
First, we view the range outputs as evaluations of f(x) over domain generator omega^n. And let selection(x)=1 for all x=omega^i where i in [0, # of elements in the range minus 1]; selection(x)=0, otherwise.
Then the high-level idea is to prove selection(x) * [f(x)+step-f(omega * x)] will always be zero for all x=omega^n over the whole domain.
Let F(x) = selection(x) * [f(x)+step-f(omega * x)] , we can turn the above problem into proving a poly t(x) exists and F(x)=t(x)V(x), where V(x) is the vanishing polynomial over the domain. To do this, the proving procedure can be divided into 2 steps:
1. Given a random z sampled by verifier, we open F, t, and V at z and check F(z)==t(z)V(z).
2. We prove the openings in step 1 are all valid.

Vimalakirti committed:
add RangeConstBB

Vimalakirti commented:
Added RangeConstBasicBlock, which allows us to skip all proving when all start, limit, and delta are known during the graph compilation.

ddkang commented on src/basic_block/range.rs:
```diff
@@ -1,26 +1,232 @@
+  pub limit: i32,
+  pub delta: i32,
+}
+impl BasicBlock for RangeConstBasicBlock {
```
How does this work for the verifier? Does the verifier need to do the MSM to prove this?

ddkang replied:
Can this be done in the setup?

Vimalakirti replied:
Added in setup

Vimalakirti committed:
verify RangeConst

Vimalakirti committed:
remove rangeBB from range layer

Vimalakirti commented:
I just realized that we can't use this RangeBasicBlock in our current implementation of zk-torch because we can't compile the graph without knowing the limit value. Otherwise, we can't infer the output shape during model compilation. But fortunately, we won't see an unknown limit in MLPerf models during model compilation. I added comments on top of RangeBasicBlock and removed it from RangeLayer.

ddkang commented:
> I just realized that we can't use this RangeBasicBlock in our current implementation of zk-torch because we can't compile the graph without knowing the limit value. Otherwise, we can't infer the output shape during model compilation. But fortunately, we won't see an unknown limit in MLPerf models during model compilation. I added comments on top of RangeBasicBlock and removed it from RangeLayer.

Can we make it so it errors out if we can't support it?

Vimalakirti commented:
> Can we make it so it errors out if we can't support it?

Yeah, we currently use RangeConstBasicBlock when all three fields are computable during graph compilation. And RangeLayer will panic when any of these fields are non-fixed.

ddkang commented on src/basic_block/range.rs:
```diff
@@ -1,26 +1,289 @@
+  pub start: i32,
+  pub delta: i32,
+}
+impl BasicBlock for RangeBasicBlock {
```
Do we need this if everything is always constant for our ONNX layers? I'd prefer to omit this for now and keep it in a branch if we can stick with constants

Vimalakirti committed:
remove rangeBB and put it in branch range_prove

Vimalakirti committed:
Merge branch 'master' into range

ddkang commented:
Looks good to me, please have lilia take a look

Vimalakirti commented:
> Looks good to me, please have lilia take a look

@liliatangxy hi, lilia, please take a look at this pr, thx!

Vimalakirti committed:
change prover and verifier cache type after gpu branch

Vimalakirti committed:
cleanup
